src/porepy/numerics/time_step_control.py

```diff
@@ -1,130 +1,191 @@
+
+        if not constant_dt:
+
+            # Sanity checks for dt_min and dt_max
+            if dt_init < dt_min_max[0]:
+                raise ValueError(
+                    "Initial time step cannot be smaller than minimum time step."
+                )
+            elif dt_init > dt_min_max[1]:
+                raise ValueError(
+                    "Initial time step cannot be larger than maximum time step."
+                )
+
+            # NOTE: The above checks guarantee that minimum time step <= maximum time step
+
+            # Sanity checks for maximum number of iterations.
+            # Note that 0 is a possibility. This will imply that the solver reaches
+            # convergence directly, e.g., as in linear solvers.
+            if iter_max < 0:
+                raise ValueError("Maximum number of iterations must be non-negative.")
+
+            # Sanity checks for optimal iteration range
+            if iter_optimal_range[0] > iter_optimal_range[1]:
+                s = "Lower optimal iteration range cannot be larger than"
+                s += " upper optimal iteration range."
+                raise ValueError(s)
+            elif iter_optimal_range[1] > iter_max:
+                s = "Upper optimal iteration range cannot be larger than"
+                s += " maximum number of iterations."
+                raise ValueError(s)
+            elif iter_optimal_range[0] < 0:
+                s = "Lower optimal iteration range cannot be negative."
+                raise ValueError(s)
+
+            # Sanity checks for lower and upper multiplication factors
+            if iter_lowupp_factor[0] <= 1:
+                raise ValueError("Expected lower multiplication factor > 1.")
+            elif iter_lowupp_factor[1] >= 1:
+                raise ValueError("Expected upper multiplication factor < 1.")
+
+            # Checks for sensible combinations of iter_optimal_range and iter_lowupp_factor
+            if dt_min_max[0] * iter_lowupp_factor[0] > dt_min_max[1]:
+                raise ValueError("Encountered dt_min * iter_low_factor > dt_max.")
+            elif dt_min_max[1] * iter_lowupp_factor[1] < dt_min_max[0]:
+                raise ValueError("Encountered dt_max * iter_upp_factor < dt_min.")
+
+            # Sanity check for recomputation factor
+            if recomp_factor >= 1:
+                raise ValueError("Expected recomputation multiplication factor < 1.")
+
+            # Sanity check for maximum number of recomputation attempts
+            if recomp_max <= 0:
+                raise ValueError("Number of recomputation attempts must be > 0.")
+
+        else:
+
+            # If the time step is constant, check that the scheduled times and the time
+            # step are compatible.
+            # E.g. dt=2 works with schedule = [0,2,4,8], but dt=3 does not.
+            sim_times = np.arange(schedule[0], schedule[-1] + dt_init, dt_init)
+            intersect = np.intersect1d(sim_times, schedule)
```

Comment:
I see, here is the answer to a question above. Please document this above.

Reply:
Done.